src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -87,26 +93,40 @@
                                           (filter dir-create?)
                                           (map #(.toPath (:path %)))))))
 
-(schema/defn retrieve-events
-  :- [(schema/one WatchKey "key") (schema/one [WatchEvent] "events")]
-  "Blocks until an event the watcher is concerned with has occured.
-  Returns the native WatchKey and WatchEvents"
+(schema/defn retrieve-events :- {WatchKey [WatchEvent]}
+  "Blocks until an event the watcher is concerned with has occured. Will then
+  poll for a new event, watiting at least `window-min` for a new event to
+  occur. Will continue polling for as long as there are new events that occur
+  within `window-min`, or the `window-max` time limit has been exceeded."
   [watcher :- (schema/protocol Watcher)]
   (let [watch-key (.take (:watch-service watcher))
-        events (.pollEvents watch-key)]
+        events (vec (.pollEvents watch-key))
+        time-limit (+ (System/currentTimeMillis) window-max)
+        events-by-key {watch-key events}]
     (.reset watch-key)
     (watch-new-directories! (map #(clojurize % (.watchable watch-key)) events) watcher)
-    [watch-key events]))
+      (loop [keys-events events-by-key]
+        (if-let [waiting-key (.poll (:watch-service watcher) window-min window-units)]
+          (let [waiting-events (vec (.pollEvents waiting-key))]
+            (watch-new-directories! (map #(clojurize % (.watchable waiting-key)) waiting-events) watcher)
+            (.reset waiting-key)
+            (if (< (System/currentTimeMillis) time-limit)
+              (recur (update keys-events waiting-key into waiting-events))
+              (update keys-events waiting-key waiting-events)))
+          keys-events))))
+
 
 (schema/defn process-events!
   "Process for side-effects any events that occured for watcher's watch-key"
   [watcher :- (schema/protocol Watcher)
-   watch-key :- WatchKey
-   orig-events :- [WatchEvent]]
-  (let [clojure-events (map #(clojurize % (.watchable watch-key)) orig-events)
+   events-by-key :- {WatchKey [WatchEvent]}]
+  (let [orig-events (flatten (vals events-by-key))
+        clojure-events (mapcat
+                         (fn [kv] (map #(clojurize % (.watchable (first kv))) (second kv)))
+                         events-by-key)
         callbacks @(:callbacks watcher)]
-    (log/info (trs "Got {0} event(s) for watched-path {1}"
-                   (count orig-events) (.watchable watch-key)))
+    (log/info (trs "Got {0} event(s) on path(s) {1}"
```

Comment:
Here's an example of the current intended output†, broken at my screen width, otherwise it would all be on one line:
```
2016-08-06 09:27:48,665 INFO  [p.t.s.w.filesystem-watch-core] Got 38 event(s) on path(s) ("/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/38.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/39.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/40.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/41.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/42.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/43.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/44.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/45.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/46.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/47.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/48.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/49.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/50.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/51.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/52.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/53.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/54.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/55.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/56.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/57.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/58.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/59.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/60.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/61.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/62.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/63.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/64.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/65.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/66.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/67.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/68.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/69.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/70.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/71.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/72.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/73.txt" "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/74.txt" "/var/folder
s/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470500862356-1130756734/window-max/75.txt")
```

Spending a bit of time playing with the logging, here's an example of just showing events per `watched-path`. I've included some of the debug output below as well:
```
2016-08-06 09:38:31,983 INFO  [p.t.s.w.filesystem-watch-core] Got 38 event(s) in directory /var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470501507670-1186788113/window-max
2016-08-06 09:38:32,087 DEBUG [p.t.s.w.filesystem-watch-core] Events:
({:path
  "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470501507670-1186788113/window-max/0.txt",
  :type :create}
 {:path
  "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470501507670-1186788113/window-max/1.txt",
  :type :create}
 {:path
  "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470501507670-1186788113/window-max/10.txt",
  :type :create}
 {:path
  "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470501507670-1186788113/window-max/11.txt",
  :type :create}
 {:path
  "/var/folders/4l/f8knljcx4pl4sh14txnm56740000gn/T/debounce1470501507670-1186788113/window-max/12.txt",
  :type :create}
```

Having played with it a bit, I think showing per watched-path is a good clean level of info for the INFO level. I'm going to push that up. Lemme know what you think!

† This actually requires a small update to also call str on the elements of the map